From 7ce0c989c8486f6f454a47e8459db149e2cf77c8 Mon Sep 17 00:00:00 2001
From: Maxim Krivchikov <maxim.krivchikov@gmail.com>
Date: Sat, 29 Aug 2020 11:24:08 +0300
Subject: [PATCH] Fix GTK+2 icons since commit 11aa6821

After 11aa6821 a lot of "missing icons" appeared in GTK+2 UI
due to using stock name in functions which expect icon name.

This commit restores icons in toolbar and in some menu items.
Before 11aa6821 menu didn't have any icons, so it's not a regression.
To add more icons to GTK+2 menu one can translate more icon names
from toolbar_items.h to NSGTK_STOCK_* in compat.h
---
 frontends/gtk/compat.h        | 4 ++++
 frontends/gtk/scaffolding.c   | 6 +++---
 frontends/gtk/toolbar.c       | 9 +++++++++
 frontends/gtk/toolbar_items.h | 4 ++--
 4 files changed, 18 insertions(+), 5 deletions(-)

diff --git a/frontends/gtk/compat.h b/frontends/gtk/compat.h
index 3b2f55094..76af0ef34 100644
--- a/frontends/gtk/compat.h
+++ b/frontends/gtk/compat.h
@@ -41,6 +41,8 @@
 #define NSGTK_STOCK_CANCEL "_Cancel"
 #define NSGTK_STOCK_CLEAR "edit-clear"
 #define NSGTK_STOCK_CLOSE "window-close"
+#define NSGTK_STOCK_GO_BACK "go-previous"
+#define NSGTK_STOCK_GO_FORWARD "go-next"
 #define NSGTK_STOCK_HOME "go-home"
 #define NSGTK_STOCK_INFO "dialog-information"
 #define NSGTK_STOCK_REFRESH "view-refresh"
@@ -55,6 +57,8 @@
 #define NSGTK_STOCK_CANCEL GTK_STOCK_CANCEL
 #define NSGTK_STOCK_CLEAR GTK_STOCK_CLEAR
 #define NSGTK_STOCK_CLOSE GTK_STOCK_CLOSE
+#define NSGTK_STOCK_GO_BACK GTK_STOCK_GO_BACK
+#define NSGTK_STOCK_GO_FORWARD GTK_STOCK_GO_FORWARD
 #define NSGTK_STOCK_HOME GTK_STOCK_HOME
 #define NSGTK_STOCK_INFO GTK_STOCK_INFO
 #define NSGTK_STOCK_REFRESH GTK_STOCK_REFRESH
diff --git a/frontends/gtk/scaffolding.c b/frontends/gtk/scaffolding.c
index f08d5d0d0..75a2f59d8 100644
--- a/frontends/gtk/scaffolding.c
+++ b/frontends/gtk/scaffolding.c
@@ -1157,17 +1157,17 @@ static void nsgtk_menu_set_icons(struct nsgtk_scaffolding *g)
 		}
 
 		if (g->menus[i].main != NULL) {
-		img = gtk_image_new_from_icon_name(g->menus[i].iconname,
+			img = nsgtk_image_new_from_stock(g->menus[i].iconname,
 						   GTK_ICON_SIZE_MENU);
 			nsgtk_image_menu_item_set_image(GTK_WIDGET(g->menus[i].main), img);
 		}
 		if (g->menus[i].burger != NULL) {
-		img = gtk_image_new_from_icon_name(g->menus[i].iconname,
+			img = nsgtk_image_new_from_stock(g->menus[i].iconname,
 						   GTK_ICON_SIZE_MENU);
 			nsgtk_image_menu_item_set_image(GTK_WIDGET(g->menus[i].burger), img);
 		}
 		if (g->menus[i].popup != NULL) {
-		img = gtk_image_new_from_icon_name(g->menus[i].iconname,
+			img = nsgtk_image_new_from_stock(g->menus[i].iconname,
 						   GTK_ICON_SIZE_MENU);
 			nsgtk_image_menu_item_set_image(GTK_WIDGET(g->menus[i].popup), img);
 		}
diff --git a/frontends/gtk/toolbar.c b/frontends/gtk/toolbar.c
index 32adabc5b..d537b254e 100644
--- a/frontends/gtk/toolbar.c
+++ b/frontends/gtk/toolbar.c
@@ -478,7 +478,11 @@ make_toolbar_item_button(const char *labelmsg,
 	}
 
 	if (item != NULL) {
+#if GTK_CHECK_VERSION(3,0,0)
 		gtk_tool_button_set_icon_name(GTK_TOOL_BUTTON(item), iconname);
+#else
+		gtk_tool_button_set_stock_id(GTK_TOOL_BUTTON(item), iconname);
+#endif
 
 		gtk_widget_set_sensitive(GTK_WIDGET(item), sensitivity);
 		if (edit) {
@@ -1205,8 +1209,13 @@ static nserror set_item_action(struct nsgtk_toolbar *tb, int itemid, bool alt)
 	gtk_tool_button_set_label(GTK_TOOL_BUTTON(tb->items[itemid].button),
 				  label);
 
+	#if GTK_CHECK_VERSION(3,0,0)
 	gtk_tool_button_set_icon_name(GTK_TOOL_BUTTON(tb->items[itemid].button),
 				      iconname);
+	#else
+	gtk_tool_button_set_stock_id(GTK_TOOL_BUTTON(tb->items[itemid].button),
+				      iconname);
+	#endif
 
 	gtk_widget_set_sensitive(GTK_WIDGET(tb->items[itemid].button), TRUE);
 
diff --git a/frontends/gtk/toolbar_items.h b/frontends/gtk/toolbar_items.h
index b4bed371f..af098788e 100644
--- a/frontends/gtk/toolbar_items.h
+++ b/frontends/gtk/toolbar_items.h
@@ -98,9 +98,9 @@ typedef enum {
 #define TOOLBAR_ITEM_SET
 #endif
 
-TOOLBAR_ITEM(BACK_BUTTON, back, false, b, p, gtkBack, "go-previous")
+TOOLBAR_ITEM(BACK_BUTTON, back, false, b, p, gtkBack, NSGTK_STOCK_GO_BACK)
 TOOLBAR_ITEM(HISTORY_BUTTON, history, true, y, n, , "local-history")
-TOOLBAR_ITEM(FORWARD_BUTTON, forward, false, b, p, gtkForward, "go-next")
+TOOLBAR_ITEM(FORWARD_BUTTON, forward, false, b, p, gtkForward, NSGTK_STOCK_GO_FORWARD)
 TOOLBAR_ITEM(STOP_BUTTON, stop, false, t, p, gtkStop, NSGTK_STOCK_STOP)
 TOOLBAR_ITEM(RELOAD_BUTTON, reload, true, t, p, Reload, NSGTK_STOCK_REFRESH)
 TOOLBAR_ITEM(HOME_BUTTON, home, true, b, p, gtkHome, NSGTK_STOCK_HOME)
-- 
2.25.4

